File: populating-next-right-pointers-in-each-node/populating-next-right-pointers-in-each-node.cs
/*
// Definition for a Node.
public class Node {
    public int val;
    public Node left;
    public Node right;
    public Node next;

    public Node() {}

    public Node(int _val) {
        val = _val;
    }

    public Node(int _val, Node _left, Node _right, Node _next) {
        val = _val;
        left = _left;
        right = _right;
        next = _next;
    }
}
*/

public class Solution {
    public Node Connect(Node root, Node next=null) {
        if(root==null) return root;
        
        root.next = next;
        
        Node nextForLeft = root.right;
        Node nextForRight = next!=null? next.left : null;
        
        Connect(root.left, nextForLeft);
        Connect(root.right, nextForRight);
        
        return root;
    }
}

File: populating-next-right-pointers-in-each-node/populating-next-right-pointers-in-each-node.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val: int = 0, left: 'Node' = None, right: 'Node' = None, next: 'Node' = None):
        self.val = val
        self.left = left
        self.right = right
        self.next = next
"""

class Solution:
    def connect(self, root: 'Optional[Node]') -> 'Optional[Node]':
        if not root:
            return root
        
        queue = [root]
        
        while len(queue):
            curr = queue.pop(0)
            Qrange = len(queue)
            
            if curr.left:
                queue.append(curr.left)
            if curr.right:
                queue.append(curr.right)
            
            for i in range(Qrange):
                curr.next = queue.pop(0)
                curr = curr.next
                
                if(curr.left):
                    queue.append(curr.left)
                if curr.right:
                    queue.append(curr.right)
        
        return root
            